java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -178,190 +232,129 @@ final boolean isSuperOrThisMemberSelect(ExpressionTree tree) {
-      if ("false".equals(tree.value())) {
-        currentResult.falseStates.add(currentState);
-      } else if ("true".equals(tree.value())) {
-        currentResult.trueStates.add(currentState);
-      } else {
-        currentResult.unknownStates.add(currentState);
-      }
-    }
-
-    @Override
-    public final void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
-      if (isSuperOrThisMemberSelect(tree)) {
-        scan(tree.identifier());
-      } else {
-        currentResult.unknownStates.add(currentState);
+        executionState.push(expression);
       }
     }
 
     @Override
-    public final void visitMethodInvocation(MethodInvocationTree tree) {
-      currentState.invalidateFields();
-      currentResult.unknownStates.add(currentState);
-    }
-
-    @Override
-    public final void visitUnaryExpression(UnaryExpressionTree tree) {
-      if (tree.is(Tree.Kind.LOGICAL_COMPLEMENT)) {
-        PackedStates unaryResult = evaluateCondition(currentState, tree.expression());
-        currentResult.falseStates.addAll(unaryResult.trueStates);
-        currentResult.trueStates.addAll(unaryResult.falseStates);
-        currentResult.unknownStates.addAll(unaryResult.unknownStates);
+    public final void visitBinaryExpression(BinaryExpressionTree tree) {
+      if (tree.is(Tree.Kind.CONDITIONAL_AND)) {
+        evaluateConditionalAnd(tree);
+      } else if (tree.is(Tree.Kind.CONDITIONAL_OR)) {
+        evaluateConditionalOr(tree);
+      } else if (tree.is(Tree.Kind.EQUAL_TO)) {
+        evaluateRelationalOperator(tree, SymbolicRelation.EQUAL_TO);
+      } else if (tree.is(Tree.Kind.GREATER_THAN)) {
+        evaluateRelationalOperator(tree, SymbolicRelation.GREATER_THAN);
+      } else if (tree.is(Tree.Kind.GREATER_THAN_OR_EQUAL_TO)) {
+        evaluateRelationalOperator(tree, SymbolicRelation.GREATER_EQUAL);
+      } else if (tree.is(Tree.Kind.LESS_THAN)) {
+        evaluateRelationalOperator(tree, SymbolicRelation.LESS_THAN);
+      } else if (tree.is(Tree.Kind.LESS_THAN_OR_EQUAL_TO)) {
+        evaluateRelationalOperator(tree, SymbolicRelation.LESS_EQUAL);
+      } else if (tree.is(Tree.Kind.NOT_EQUAL_TO)) {
+        evaluateRelationalOperator(tree, SymbolicRelation.NOT_EQUAL);
       } else {
-        currentResult.unknownStates.add(currentState);
-      }
-    }
-
-    @Override
-    void evaluateConditionalAnd(BinaryExpressionTree tree) {
-      PackedStates leftResult = evaluateCondition(currentState, tree.leftOperand());
-      currentResult.falseStates.addAll(leftResult.falseStates);
-      for (ExecutionState state : leftResult.trueStates) {
-        PackedStates rightResult = evaluateCondition(state, tree.rightOperand());
-        currentResult.add(rightResult);
+        super.visitBinaryExpression(tree);
+        for (ExecutionState executionState : currentStates) {
+          executionState.pop();
+          executionState.pop();
+          executionState.push(new SymbolicValue());
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this instantiation of "SymbolicValue". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)